edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.CloudProxy/authenticators/DeviceScopeTokenAuthenticator.cs

```diff
@@ -19,64 +19,40 @@ public class DeviceScopeTokenAuthenticator : IAuthenticator
             this.iothubHostName = Preconditions.CheckNonWhiteSpace(iothubHostName, nameof(iothubHostName));
             this.edgeHubHostName = Preconditions.CheckNotNull(edgeHubHostName, nameof(edgeHubHostName));
-            this.connectionManager = Preconditions.CheckNotNull(connectionManager, nameof(connectionManager));
-            this.deviceScopeIdentitiesCache.ServiceIdentityUpdated += this.HandleServiceIdentityUpdate;
-            this.deviceScopeIdentitiesCache.ServiceIdentityRemoved += this.HandleServiceIdentityRemove;
-        }
+        }              
 
-        async void HandleServiceIdentityUpdate(object sender, ServiceIdentity serviceIdentity)
+        public async Task<bool> AuthenticateAsync(IClientCredentials clientCredentials)
         {
-            try
-            {
-                Events.ServiceIdentityUpdated(serviceIdentity.Id);
-                Option<IClientCredentials> clientCredentials = this.connectionManager.GetClientCredentials(serviceIdentity.Id);
-                await clientCredentials.ForEachAsync(
-                    async c =>
-                    {
-                        if (!(c is ITokenCredentials tokenCredentials) ||
-                            !await this.AuthenticateInternalAsync(tokenCredentials, serviceIdentity))
-                        {
-                            Events.ServiceIdentityUpdatedRemoving(serviceIdentity.Id);
-                            await this.connectionManager.RemoveDeviceConnection(c.Identity.Id);
-                        }
-                        else
-                        {
-                            Events.ServiceIdentityUpdatedValidated(serviceIdentity.Id);
-                        }
-                    });
-            }
-            catch (Exception ex)
+            if (!(clientCredentials is ITokenCredentials tokenCredentials))
             {
-                Events.ErrorReauthenticating(ex, serviceIdentity);
+                return false;
             }
-        }
 
-        async void HandleServiceIdentityRemove(object sender, string id)
-        {
             try
             {
-                Events.ServiceIdentityRemoved(id);
-                await this.connectionManager.RemoveDeviceConnection(id);
+                Option<ServiceIdentity> serviceIdentity = await this.deviceScopeIdentitiesCache.GetServiceIdentity(clientCredentials.Identity.Id);
+                return await serviceIdentity.Map(s => this.AuthenticateInternalAsync(tokenCredentials, s))
+                    .GetOrElse(() => this.underlyingAuthenticator.AuthenticateAsync(clientCredentials));
             }
-            catch (Exception ex)
+            catch (Exception e)
             {
-                Events.ErrorRemovingConnection(ex, id);
+                Events.ErrorAuthenticating(e, clientCredentials);
+                return await this.underlyingAuthenticator.AuthenticateAsync(clientCredentials);
             }
```

Comment:
Can this be called only when AuthenticateInternalAsync fails, no need to call it if the underlyingAuthenticator failed?

Reply:
Good point, will fix.